cmd/flat-network-controller/app/controller.go

```diff
@@ -0,0 +1,247 @@
+	for i := 0; i < workers; i++ {
+		go wait.Until(c.runWorker, time.Second, stopCh)
+	}
+
+	klog.V(5).Infof("%d workers started", workers)
+	<-stopCh
+	klog.Info("shutting down flat network controller")
+}
+
+func (c *Controller) runWorker() {
+	for {
+		item, shutdown := c.queue.Get()
+		if shutdown {
+			break
+		}
+
+		c.process(item)
+	}
+}
+
+// processItem will read a single work item off the work queue and attempt to process it
+func (c *Controller) process(item interface{}) {
+	defer c.queue.Done(item)
+
+	key, ok := item.(string)
+	if !ok {
+		klog.Errorf("unexpected item in queue: %v", item)
+		c.queue.Forget(item)
+		return
+	}
+
+	tenant, name, err := parseKey(key)
+	if err != nil {
+		klog.Errorf("unexpected string in queue; discarding: %s", key)
+		c.queue.Forget(item)
+		return
+	}
+
+	net, err := c.store.NetworksWithMultiTenancy(tenant).Get(name)
+	if err != nil {
+		klog.Warningf("failed to retrieve network in local cache by tenant %s, name %s: %v", tenant, name, err)
+		c.queue.Forget(item)
+		return
+	}
+
+	if net.Spec.Type != "flat" {
+		klog.V(5).Infof("network %s/%s is of type %q; ignored", net.Tenant, net.Name, net.Spec.Type)
+		c.queue.Forget(item)
+		return
+	}
+
+	klog.V(5).Infof("processing network %s/%s", net.Tenant, net.Name)
+
+	if err := manageFlatNetwork(net, c.netClientset, c.svcClientset); err != nil {
+		c.queue.AddRateLimited(key)
+		c.recorder.Eventf(net, corev1.EventTypeWarning, "FailedProvision", "failed to provision network %s/%s/%s: %v", net.Tenant, net.Tenant, net.Name, err)
+		return
+	}
+
+	c.recorder.Eventf(net, corev1.EventTypeNormal, "SuccessfulProvision", "successfully provision network %s/%s/%s", net.Tenant, net.Tenant, net.Name)
+	c.queue.Forget(item)
+}
+
+// manageFlatNetwork is the core logic to manage a flat typed network object
+func manageFlatNetwork(net *v1.Network, netClient arktos.Interface, svcClient kubernetes.Interface) error {
+	if len(net.Status.DNSServiceIP) != 0 {
+		klog.V(5).Infof("network %s/%s/%s already have DNS service IP %s; skipped", net.Tenant, net.Namespace, net.Name, net.Status.DNSServiceIP)
+		return nil
+	}
+
+	svc, err := createOrGetDNSService(net, svcClient)
+	if err != nil {
+		return err
+	}
+
+	// since dns service is in place, flat type network is Ready now
+	netReady := net.DeepCopy()
+	netReady.Status.Phase = v1.NetworkReady
+	netReady.Status.DNSServiceIP = svc.Spec.ClusterIP
+	netReady.Status.Message = "DNS service ready; network ready"
+	_, err = netClient.ArktosV1().NetworksWithMultiTenancy(netReady.Tenant).UpdateStatus(netReady)
+	return err
+}
+
+func createOrGetDNSService(net *v1.Network, svcClient kubernetes.Interface) (*corev1.Service, error) {
+	const nsDNS = "kube-system"
+	nameDNS := "kube-dns-" + net.Name
+	dns := corev1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      nameDNS,
+			Tenant:    net.Tenant,
+			Namespace: nsDNS,
+		},
+		Spec: corev1.ServiceSpec{
+			Ports: []corev1.ServicePort{
+				{
+					Name:       "dns",
```

Comment:
"dns" -> "dns-udp"?

Reply:
"dns" seems the proper name following the suit of k8s dns service.